tokio/src/sync/mpsc/chan.rs

```diff
@@ -293,6 +297,79 @@ impl<T, S: Semaphore> Rx<T, S> {
         })
     }
 
+    /// Receives available values into result buffer whose capacity increases as needed
+    pub(crate) fn recv_many(&mut self, cx: &mut Context<'_>, buffer: &mut Vec<T>) -> Poll<usize> {
+        use super::block::Read::*;
+
+        ready!(crate::trace::trace_leaf(cx));
+
+        // Keep track of task budget
+        let coop = ready!(crate::runtime::coop::poll_proceed(cx));
+
+        buffer.clear();
+
+        self.inner.rx_fields.with_mut(|rx_fields_ptr| {
+            let rx_fields = unsafe { &mut *rx_fields_ptr };
+
+            macro_rules! try_recv {
+                () => {
+                    match rx_fields.list.pop(&self.inner.tx) {
+                        Some(Value(value)) => {
+                            let capacity = self.inner.semaphore.num_acquired();
+                            if buffer.capacity() < capacity {
+                                buffer.reserve(capacity);
+                            }
+                            buffer.push(value);
+                            let mut next = rx_fields.list.peek();
+                            while (buffer.len() < capacity
+                                && match next {
+                                    Some(Value(value)) => {
+                                        rx_fields.list.pop(&self.inner.tx);
+                                        buffer.push(value);
+                                        next = rx_fields.list.peek();
+                                        true
+                                    }
+                                    _ => false,
+                                })
+                            {}
+                            self.inner.semaphore.add_permits(buffer.len());
+                            coop.made_progress();
+                            return Ready(buffer.len());
```

Comment:
It is not obvious to me that this code is correct. Would it be possible to simplify it?

Reply:
simplified when reworked.